verilog/CST/package_test.cc

```diff
@@ -91,6 +101,35 @@ endclass
   EXPECT_EQ(package_declarations.size(), 2);
 }
 
+TEST(FindAllPackageDeclarationsTest2, MultiPackages2) {
+  //constexpr int kTag = 1;
+  const SyntaxTreeSearchTestCase testcases[] = {
+       {"package p;\n",
+       "endpackage\n"},
+       {"package mod; endpackage"}
+
+  };
+
+    const auto & test = testcases[1];
+    const absl::string_view code(test.code);
+    VerilogAnalyzer analyzer(code, "test-file");
+    //const auto code_copy = analyzer.Data().Contents();
+    ASSERT_OK(analyzer.Analyze()) << "failed on:\n" << code;
+    const auto& root = analyzer.Data().SyntaxTree();
+
+    const auto decls = FindAllPackageDeclarations(*root);
+
+    //EXPECT_EQ(decls.size(), 1);
+    //std::vector<TreeSearchMatch> types;
+    const auto& package_node =
+    down_cast<const SyntaxTreeNode&>(*decls.front().match);
+    const auto& token = GetPackageNameToken(package_node);
+    EXPECT_EQ(token.text(), "mod");
+    //std::ostringstream diffs;
```

Comment:
Take a look at the tests in https://github.com/google/verible/blob/master/verilog/CST/declaration_test.cc

You can write test cases that embed expected tokens that way.  `ExactMatchFindings` will do all of the element comparison work for you.  It also work when you expect no matches.  (Trying breaking some tests in that file to see how it works.)

Reply:
ok have been playing with it, have somemore testcases.
and package seems to be always tagged in pairs with endpackage, is there a way to split them up into below test :

package p;
//some code

endpackage

or is this where getpackagenametoken comes in?

Reply:
You can find more package declaration test cases and examples here (search for endpackage):

https://github.com/google/verible/blob/master/verilog/parser/verilog_parser_unittest.cc

And yes, many language elements are paired with `end*` keywords.

Reply:
thanks so much for the help! Have added more testcases and removed the commented code. Are these good to move to next task or anymore change needed?

Reply:
Last suggestion for this exercise, as it pertains to the original task of finding declared identifiers:
Create or modify tests that pick out the declared package name identifiers, `{"package ", {kTag, "my_package"}, ...}`
(You may also replace/delete older tests that are more than covered by your new one.)

Reply:
ok noted, i have sample test like this below : 

{"package ", {kTag, "foo"}, "; \n endpackage",
      "package ", {kTag, "bar"}, "; \n endpackage"}


However, i am struggling to fix below, do i need to perform some casting? since GetPackageNameToken returns TokenInfo, instead of symbol that is expected in TreeSearchMatch object

const auto declarations = FindAllPackageDeclarations(*ABSL_DIE_IF_NULL(root));

    std::vector<TreeSearchMatch> declIdentifiers;
    for (const auto& decl : declarations) {
      const auto& packageToken = GetPackageNameToken(*decl.match);
      declIdentifiers.push_back(TreeSearchMatch{&packageToken, {}});
    }

Reply:
Not casting but, you may need to have `GetPackageNameToken` return `SyntaxTreeLeaf` instead of the `TokenInfo` inside it.

Reply:
ok i made another function getpackagenametoken2 to test it first,  since getpackagenametoken affects some code in verilog/analysis/checkers/package_filename_rule* 

also just placed a couple testcases for now, if the code looks good, il add more testcases during the weekend